### PR TITLE
Add support for client certificate

### PR DESCRIPTION
**Pull request by r3no1t**
did cleanup the PR and did a forced push, now only the commit with the client certificate support is added including the review findings

### CODE REVIEW

File: src/tcp_mbedtls.cpp [hunk 1]
```diff
@@ -185,6 +187,8 @@ tcp_ssl_t * tcp_ssl_new(struct tcp_pcb *tcp, void* arg) {
   new_item->tcp_pbuf = NULL;
   new_item->pbuf_offset = 0;
   new_item->next = NULL;
+  new_item->has_ca_cert = NULL;
+  new_item->has_client_cert = NULL;
```
Comment (tve) [inline src/tcp_mbedtls.cpp]: Use false or 0, not NULL for a bool... (I know, NULL==0 but it's confusing)

File: src/tcp_mbedtls.cpp [hunk 2]
```diff
@@ -263,6 +278,21 @@ int tcp_ssl_new_client(struct tcp_pcb *tcp, void *arg, const char* hostname, con
     mbedtls_ssl_conf_authmode(&tcp_ssl->ssl_conf, MBEDTLS_SSL_VERIFY_OPTIONAL);
   }
 
+  if (tcp_ssl->has_client_cert) {
+    TCP_SSL_DEBUG("Loading CRT cert");
+    ret = mbedtls_x509_crt_parse(&tcp_ssl->client_cert, (const unsigned char *) cli_cert, cli_cert_len);
+    if (ret < 0) {
+      tcp_ssl_free(tcp);
+      return handle_error(ret);
+    } TCP_SSL_DEBUG("Loading private key");
```
Comment (tve) [inline src/tcp_mbedtls.cpp]: missing newline

**Review by tve — CHANGES_REQUESTED** [review]
Good stuff! Two main issues:
- please make the PR against my mbed-tls-try2 branch, not master
- please revert 5755043cde55c13f8e4811f5bb1aada874345dfd in order to leave the #if ASYNC_TCL_SSL_ENABLED in the code: it saves a lot of code space for those that don't want SSL...

**Comment by robert-alfaro** [comment]
Is there a need to convert tcp_mbedtls.c to .cpp? Keep the externs and favor C.

**Comment by tve** [comment]
nice! how did you test?

**Comment by r3no1t** [comment]
I'm using it in the esphome device framework, have around 30 wifi nodes connected to an MQTT broker where I use client and server authentication.

**Comment by tve** [comment]
I hadn't noticed, but github says "This branch has conflicts that must be resolved" -- can you resolve them? Then I'd be happy to merge :-)

**Comment by r3no1t** [comment]
sorry, my fork was not rebased, should now be ok

**Comment by tve** [comment]
Thanks much!

**Comment by mrdc** [comment]
@r3no1t 

Hi!
Will you be so kind to provide an example on how to set the client certificate & key?
`AsyncMqttClient` has only `setPsk` method, but not `setClientCert` or `setClientKey`, so it not clear how to pass it from   `AsyncClient` to `AsyncMqttClient`.

**Comment by r3no1t** [comment]
@mrdc the changes on the async-mqtt-client repo are not yet merged back, so you would need the branch: https://github.com/r3no1t/async-mqtt-client/tree/add_client_certificate_support

**Comment by mrdc** [comment]
@r3no1t 

Oh, thanks! I see that changes were not merged with tve's asyncmqtt repo, that's why methods are missing.

**Comment by mrdc** [comment]
@r3no1t 

Unfortunately, compilation fails. I've also tried your AsyncTCP branch which was merged to tve's fork.
```
https://github.com/r3no1t/async-mqtt-client.git#add_client_certificate_support
https://github.com/r3no1t/AsyncTCP.git#add-support-for-client-certificate
```

```
.pio/libdeps/AsyncMqttClient-esphome/src/AsyncMqttClient.cpp:204:5: error: 'SSL' was not declared in this scope
     SSL* clientSsl = _client.getSSL();
     ^
.pio/libdeps/AsyncMqttClient-esphome/src/AsyncMqttClient.cpp:204:10: error: 'clientSsl' was not declared in this scope
     SSL* clientSsl = _client.getSSL();
          ^
.pio/libdeps/AsyncMqttClient-esphome/src/AsyncMqttClient.cpp:204:30: error: 'class AsyncClient' has no member named 'getSSL'
     SSL* clientSsl = _client.getSSL();
                              ^
.pio/libdeps/AsyncMqttClient-esphome/src/AsyncMqttClient.cpp:208:62: error: 'ssl_match_fingerprint' was not declared in this scope
       if (ssl_match_fingerprint(clientSsl, fingerprint.data()) == SSL_OK) {
                                                              ^
.pio/libdeps/AsyncMqttClient-esphome/src/AsyncMqttClient.cpp:208:67: error: 'SSL_OK' was not declared in this scope
       if (ssl_match_fingerprint(clientSsl, fingerprint.data()) == SSL_OK) {
```
Will check later - maybe it's because of some mess with dependencies.

**Comment by mrdc** [comment]
Fix for this error when compiling for ESP32:
~~`#if defined (ASYNC_TCP_SSL_ENABLED) && ! defined (ASYNC_TCP_CERTIFICATE_ENABLED) && !defined(ESP32)`
https://github.com/r3no1t/async-mqtt-client/blob/178fbdb2fcd7e0ca911da04f4cbe04c9fe7ba03b/src/AsyncMqttClient.cpp#L202~~

**Comment by r3no1t** [comment]
you just need to use these two defines when building ASYNC_TCP_SSL_ENABLED ASYNC_TCP_CERTIFICATE_ENABLED

**Comment by mrdc** [comment]
@r3no1t 

Yes, the code for server fingerprint check shouldn't be used when `ASYNC_TCP_CERTIFICATE_ENABLED`, but  in my case it fails. I'll check it.

Everything is fine now: compiles and all methods are accessible.

**Comment by mrdc** [comment]
@r3no1t 

Maybe you know how to enable mbedTLS debug output?
With
```
#if 1
#define TCP_SSL_DEBUG(...) do { ets_printf("T %s- ", pcTaskGetTaskName(xTaskGetCurrentTaskHandle())); ets_printf(__VA_ARGS__); } while(0)
```
and 
```
build_flags = 
    -DCORE_DEBUG_LEVEL=5
```
I don't see any info related to tls...

**Pull request merged**